phoenix-core/src/it/java/org/apache/phoenix/replication/ReplicationLogGroupIT.java

```diff
@@ -0,0 +1,339 @@
+                    "constraint pk primary key (id1, id2))", tableName);
+            conn.createStatement().execute(ddl);
+            ddl = String.format("create index %s on %s (val1) include (val2)",
+                    indexName1, tableName);
+            conn.createStatement().execute(ddl);
+            ddl = String.format("create index %s on %s (val2) include (val1)",
+                    indexName2, tableName);
+            conn.createStatement().execute(ddl);
+            ddl = String.format("create local index %s on %s (id2,val1) include (val2)",
+                    indexName3, tableName);
+            conn.createStatement().execute(ddl);
+            conn.commit();
+            PreparedStatement stmt = conn.prepareStatement(
+                    "upsert into " + tableName + " VALUES(?, ?, ?, ?)");
+            // upsert 50 rows
+            int rowCount = 50;
+            for (int i = 0; i < 5; ++i) {
+                for (int j = 0; j < 10; ++j) {
+                    stmt.setInt(1, i);
+                    stmt.setInt(2, j);
+                    stmt.setString(3, "abcdefghijklmnopqrstuvwxyz");
+                    stmt.setString(4, null);
+                    stmt.executeUpdate();
+                }
+                conn.commit();
+            }
+            // do some atomic upserts which will be ignored and therefore not replicated
+            stmt = conn.prepareStatement("upsert into " + tableName + " VALUES(?, ?, ?) " +
+                    "ON DUPLICATE KEY IGNORE");
+            conn.setAutoCommit(true);
+            for (int i = 0; i < 5; ++i) {
+                for (int j = 0; j < 2; ++j) {
+                    stmt.setInt(1, i);
+                    stmt.setInt(2, j);
+                    stmt.setString(3, null);
+                    assertEquals(0, stmt.executeUpdate());
+                }
+            }
+            // verify the correctness of the index
+            IndexToolIT.verifyIndexTable(tableName, indexName1, conn);
+            // verify replication
+            Map<String, Integer> expected = Maps.newHashMap();
+            // mutation count will be equal to row count since the atomic upsert mutations will be
+            // ignored and therefore not replicated
+            expected.put(tableName, rowCount * 3); // Put + Delete + local index update
+            // for index1 unverified + verified + delete (Delete column)
+            expected.put(indexName1, rowCount * 3);
+            // for index2 unverified + verified  since the null column is part of row key
+            expected.put(indexName2, rowCount * 2);
+            // we didn't create any tenant views so no change in the syscat entries
+            expected.put(SYSTEM_CATALOG_NAME, 0);
+            expected.put(SYSTEM_CHILD_LINK_NAME, 0);
+            verifyReplication(conn, expected);
+        }
+    }
+
+    @Test
+    public void testPreWALRestoreSkip() throws Exception {
```

Comment:
This test name should be testWALRestore, I think. We should add a high level description of what this test does, something like this test simulates region server crashes in the middle of write transactions.